Unique transaction names for Swagger

honzajavorek opened the pull request:
**BREAKING CHANGE**

This PR addresses https://github.com/apiaryio/dredd/issues/558

### Background

Dredd uses so called _transaction names_ (the `API Name > Group Name > ... > ...` string) to address particular transactions from API description. User can write hooks, which listen on these transaction names, and modify Dredd's behavior.

Currently, transaction names have many problems and limitations and we want to move away from them (see https://github.com/apiaryio/dredd/issues/227), but that's a long run. One of the limitations is that it doesn't take into account individual requests, content types, and status codes, which makes some combinations of requests-responses impossible to address in hooks. That's a fact we live with, at the moment.

When we were adding Swagger support to Dredd, I sticked to current Dredd capabilities when designing transaction names for Swagger, so the resulting implementation has the same flaws. However, in real world use, these limitations proved to be far more damaging for Swagger users, because Swagger basically just provides the URI and HTTP method and there are many duplicate transaction names as a result.

### Solution

We decided to make a step forward https://github.com/apiaryio/dredd/issues/227 and to add status code and content type as part of transaction names generated for Swagger documents. This change is attached to the existing code by duct tape and leverages `exampleName` to inject the information (it was previously unused for Swagger documents). This should be treated as known increase of technical debt and should be removed once Dredd fully migrates to https://github.com/apiaryio/dredd/issues/227 in the future. Meanwhile, this change radically improves Swagger users lives.

### Breaking Change

All Swagger users of Dredd hooks will have to change how they address transactions - e.g. `/112 > GET` will have to be changed to `/112 > GET > 200 > application/json`, etc. That's why this change is breaking. The breaking change status will be propagated also to Dredd project itself once upgrading Dredd Transactions in its `package.json`.

miiila commented on src/compile.coffee:
```diff
@@ -222,4 +218,26 @@ compilePathOrigin = (filename, parseResult, httpTransaction) ->
+  transition = parent(httpTransaction, parseResult, {element: 'transition'})
+  httpResponse = child(httpTransaction, {element: 'httpResponse'})
+
+  if mediaType is 'text/vnd.apiblueprint'
```
I'd move `exampleName = ''` just above the if. No big deal, but it would help me in orientation.

honzajavorek replied:
Addressed in 4ea71db.

miiila commented on test/integration/compile-swagger-test.coffee:
```diff
@@ -97,25 +97,44 @@ describe('compile() · Swagger', ->
+      do (statusCode, i) ->
+        context("origin of transaction ##{i + 1}", ->
+          it('uses URI as resource name', ->
+            assert.equal(transactions[i].origin.resourceName, '/honey')
```
Because of `?` on line 111, transactions can be `undefined`. However it would still fail the test, I prefer to have assert fail rather than runtime error. YMMV.

miiila replied:
I prefer to have fixtures in variables rather than hardcoded, but... YMMV ;-)

honzajavorek replied:
`transactions` should™ be `undefined` only in case `compileFixture` passes an error to the callback. In that case, the `transactions = compilationResult?.transactions` line will pass, but subsequent `done(err)` line will propagate the error and stop the test. It won't ever get to line 127 in that case.

honzajavorek replied:
I made the change because previously, the `transaction = compilationResult.transactions[0]` would blow up and stop the test by crash - not by `done(err)` as intended.

miiila commented on src/compile.coffee:
```diff
@@ -184,25 +185,20 @@ compileHeaders = (httpHeaders) ->
 
 
-compileOrigin = (filename, parseResult, httpTransaction) ->
+compileOrigin = (filename, parseResult, mediaType, httpTransaction) ->
```
I see `compileOrigin` as a descendant of `compile` function, so I would vote for same signature (i.e. `(mediaType, parseResult, filename)`).

honzajavorek replied:
Addressed in 902414b.

miiila commented on test/integration/compile-swagger-test.coffee:
```diff
@@ -97,25 +97,44 @@ describe('compile() · Swagger', ->
-    compilationResult = undefined
-    transaction = undefined
+
+    expectedStatusCodes = [200, 400, 500]
```
I would change the naming in way it supports the fact you are iterating over it and evaluating all of them.

honzajavorek replied:
I need help with this. Do you have any suggestions on a better name?

miiila commented on test/integration/compile-swagger-test.coffee:
```diff
@@ -97,25 +97,44 @@ describe('compile() · Swagger', ->
+            assert.equal(transactions[i].origin.resourceName, '/honey')
+          )
+          it('uses method as action name', ->
+            assert.equal(transactions[i].origin.actionName, 'GET')
```
I prefer to have fixtures in variables rather than hardcoded, but... YMMV ;-)

honzajavorek replied:
I use literals everywhere unless it goes against DRY or the ability to parametrize the test. When reading the test, I like that literals are immediately obvious for readers and they don't have to look up the value, so more their brain cycles are left for understanding the meaning of the test.

miiila replied:
Although I understand your point, I would dispute it with a fact that reviewer is unable to verify expected data unless he would:

1. Find where the fixtures is loaded
2. Find the source of the fixtures
3. Goes to the fixture file to verify the literal

Declaring variable with expected value near the place, where the fixture is required would omit first two steps and saves more brain cycles, IMHO :-)

With current setup, I have to convince myself to make a journey to verify the expected value, which could lead to skipping typos (`goat` vs `boat`; both can be legit) in false positive test.

honzajavorek replied:
I don't see how just having a variable helps. The best would be having a
variable which is used to build the fixture and then expected in assert.
That's less easy to write, but IMHO makes the best sense and makes the test
better in every aspect. However, just moving a literal to a variable, which
still isn't tied to the actual fixture in any way, is probably just a
matter of preference.

On 8 Sep 2016 2:41 p.m., "Mila Votradovec" <notifications@github.com> wrote:

> In test/integration/compile-swagger-test.coffee
> <https://github.com/apiaryio/dredd-transactions/pull/63#discussion_r77997366>
> :
>
> >          done(err)
> >        )
> >      )
> >
> >      it('detection of transaction examples was not called', ->
> >        assert.isFalse(detectTransactionExamples.called)
> >      )
> > +    it('expected number of transactions was returned', ->
> > +      assert.equal(transactions.length, expectedStatusCodes.length)
> > +    )
> > +
> > +    for statusCode, i in expectedStatusCodes
> > +      do (statusCode, i) ->
> > +        context("origin of transaction ##{i + 1}", ->
> > +          it('uses URI as resource name', ->
> > +            assert.equal(transactions[i].origin.resourceName, '/honey')
> > +          )
> > +          it('uses method as action name', ->
> > +            assert.equal(transactions[i].origin.actionName, 'GET')
>
> Although I understand your point, I would dispute it with a fact that
> reviewer is unable to verify expected data unless he would:
>
>    1. Find where the fixtures is loaded
>    2. Find the source of the fixtures
>    3. Goes to the fixture file to verify the literal
>
> Declaring variable with expected value near the place, where the fixture
> is required would omit first two steps and saves more brain cycles, IMHO :-)
>
> With current setup, I have to convince myself to make a journey to verify
> the expected value, which could lead to skipping typos (goat vs boat;
> both can be legit) in false positive test.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/apiaryio/dredd-transactions/pull/63/files/a1c7007fd63b348eff31086d0866d2ad5538cd38#r77997366>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARTMdrwo_-sQx8fSwL7qzP-jDo3uNcMks5qoAKOgaJpZM4Js3K4>
> .
>

miiila commented on test/unit/transaction-name/get-transaction-name-test.coffee:
```diff
@@ -5,3 +5,36 @@ getTransactionName = require '../../../src/transaction-name/get-transaction-name
+        actionName: 'd'
+        exampleName: 'e > f'
+    )
+    assert.equal(name, 'a > b > c > d > e > f')
```
I would add test for single value (to verify, it doesn't contain `>` char).

honzajavorek replied:
Addressed in fa2b210.

miiila commented:
@honzajavorek Review done, nothing blocking found 👍. 
Deploy is waiting for you in order to let you decide whether you want to include comments 🚀.

honzajavorek commented:
@miiila Addressed your comments. Feel free to continue commenting if you feel something isn't all right, I'll issue fixes in subsequent PRs if necessary. Merging this now.
